lib/internal/modules/esm/utils.js

```diff
@@ -165,20 +165,41 @@ async function importModuleDynamicallyCallback(symbol, specifier, assertions) {
   throw new ERR_VM_DYNAMIC_IMPORT_CALLBACK_MISSING();
 }
 
+async function importModuleDynamicallyInShadowRealm(symbol, specifier, assertions) {
+  if (moduleRegistries.has(symbol)) {
+    const { importModuleDynamically, callbackReferrer } = moduleRegistries.get(symbol);
+    if (importModuleDynamically !== undefined) {
+      return importModuleDynamically(specifier, callbackReferrer, assertions);
+    }
+  }
+
+  // The handler for `ShadowRealm.prototype.importValue`.
+  const { esmLoader } = require('internal/process/esm_loader');
+  // `parentURL` is not set in the case of a ShadowRealm top-level import.
+  return esmLoader.import(specifier, undefined, assertions);
+}
+
 let _isLoaderWorker = false;
 /**
  * Initializes handling of ES modules.
  * This is configured during pre-execution. Specifically it's set to true for
  * the loader worker in internal/main/worker_thread.js.
  * @param {boolean} [isLoaderWorker=false] - A boolean indicating whether the loader is a worker or not.
+ * @param {boolean} [isShadowRealm=false] - A boolean indicating whether the loader is a ShadowRealm.
  */
-function initializeESM(isLoaderWorker = false) {
+function initializeESM(isLoaderWorker = false, isShadowRealm = false) {
   _isLoaderWorker = isLoaderWorker;
```

Comment:
I think we can just update this to `_forceDefaultLoader` and rename the argument so it can be used for shadow realms too. The `isShadowRealm` is essentially doing the same thing as `isLoaderWorker`, there's no need to invent a new way to bypass the custom loader.